test/jdk/jdk/internal/util/ByteArray/Types.java

```diff
@@ -0,0 +1,184 @@
+        UnaryOperator<T> reverse = in.reverse;
+        Comparator<T> comparator = in.comparator;
+        int arrayLen = 128;
+        byte[] arr = new byte[arrayLen];
+
+        assertThrows(NullPointerException.class, () -> orderedReader.get(null, 0, true));
+        assertThrows(NullPointerException.class, () -> beReader.get(null, 0));
+        assertThrows(NullPointerException.class, () -> leReader.get(null, 0));
+        assertThrows(IndexOutOfBoundsException.class, () -> orderedReader.get(arr, -1, false));
+        assertThrows(IndexOutOfBoundsException.class, () -> beReader.get(arr, -1));
+        assertThrows(IndexOutOfBoundsException.class, () -> leReader.get(arr, -1));
+        assertThrows(IndexOutOfBoundsException.class, () -> orderedReader.get(arr, arrayLen - size + 1, false));
+        assertThrows(IndexOutOfBoundsException.class, () -> beReader.get(arr, arrayLen - size + 1));
+        assertThrows(IndexOutOfBoundsException.class, () -> leReader.get(arr, arrayLen - size + 1));
+
+        Random random = RandomFactory.getRandom();
+        random.nextBytes(arr);
+        for (int index = 0; index + size <= arrayLen; index += random.nextInt(1, 33)) {
+            T be = beReader.get(arr, index);
+            T le = leReader.get(arr, index);
+            T beO = orderedReader.get(arr, index, true);
+            T leO = orderedReader.get(arr, index, false);
+            assertEquals(0, comparator.compare(be, beO));
+            assertEquals(0, comparator.compare(le, leO));
+            if (reverse != null) {
+                assertEquals(0, comparator.compare(reverse.apply(be), le));
+            }
+        }
+    }
+
+    interface OrderedWriter<T> {
+        void set(byte[] array, int index, boolean big, T value);
+    }
+
+    interface Writer<T> {
+        void set(byte[] array, int index, T value);
+    }
+
+    record WriteCase<T>(String name, OrderedWriter<T> orderedWriter, Writer<T> beWriter, Writer<T> leWriter, int bytesCount, List<T> equalValues) {}
+
+    static Stream<WriteCase<?>> writeCases() {
+        return Stream.of(
+                new WriteCase<>("char", ByteArray::setCharBO, ByteArray::setCharBE, ByteArray::setCharLE, Character.BYTES, List.of('e')),
+                new WriteCase<>("short", ByteArray::setShortBO, ByteArray::setShortBE, ByteArray::setShortLE, Short.BYTES, List.of((short) 56)),
+                new WriteCase<>("u2", ByteArray::setUnsignedShortBO, ByteArray::setUnsignedShortBE, ByteArray::setUnsignedShortLE, 2, List.of(32768, -32768)),
+                new WriteCase<>("int", ByteArray::setIntBO, ByteArray::setIntBE, ByteArray::setIntLE, Integer.BYTES, List.of(42)),
+                new WriteCase<>("float", ByteArray::setFloatBO, ByteArray::setFloatBE, ByteArray::setFloatLE, Float.BYTES, List.of(Float.NaN, Float.intBitsToFloat(0x7FF23847))),
+                new WriteCase<>("float raw", ByteArray::setFloatRawBO, ByteArray::setFloatRawBE, ByteArray::setFloatRawLE, Float.BYTES, List.of(1.0F)),
```

Comment:
Raw seems to be exercised only on unproblematic cases, not on NaNs.

Similarly for "double raw".